module: add note when cjs loader throws error

gntem opened the pull request:
subsystem: module: add note to error when import,export is detected.

These will allow users to know how to change their project to support 
es modules when the message is displayed.

##### Checklist
<!-- Remove items that do not apply. For completed items, change [ ] to [x]. -->

- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

yanivfriedensohn committed:
src: add error codes to errors thrown in C++

PR-URL: https://github.com/nodejs/node/pull/27700
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Joyee Cheung <joyeec9h3@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

tniessen committed:
crypto: extend RSA-OAEP support with oaepHash

This adds an oaepHash option to asymmetric encryption which allows
users to specify a hash function when using OAEP padding. This
feature is required for interoperability with WebCrypto applications.

PR-URL: https://github.com/nodejs/node/pull/28335
Fixes: https://github.com/nodejs/node/issues/25756
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Sam Roberts <vieuxtech@gmail.com>

danbev committed:
tools: allow single JS file for --link-module

The description for the --link-module configuration option is as
follows:
$ ./configure --help | grep -A 5 'link-module'
  --link-module=LINKED_MODULE
                      Path to a JS file to be bundled in the binary as a
                      builtin. This module will be referenced by path
                      without extension; e.g. /root/x/y.js will be
                      referenced via require('root/x/y'). Can be used
                      multiple times

This lead me to think that it was possible to specify a file like this:
$ ./configure --link-module=something.js
$ NODE_DEBUG=mkcodecache make -j8

This will lead to a compilation error as an entry in the source_ map in
node_javascript.cc will end up having an empty string as its key:
source_.emplace("", UnionBytes{_raw, 105});

This will then be used by CodeCacheBuilder when it iterates over the
module ids, which will lead to the following compilation errors:

/node/out/Release/obj/gen/node_code_cache.cc:12:23: warning:
ISO C++17 does not allow a decomposition group to be
empty [-Wempty-decomposition]
static const uint8_t [] = {
                      ^
/node/out/Release/obj/gen/node_code_cache.cc:12:22: warning:
decomposition declarations are a C++17 extension [-Wc++17-extensions]
static const uint8_t [] = {
                     ^~
/node/out/Release/obj/gen/node_code_cache.cc:12:1: error:
decomposition declaration cannot be declared 'static'
static const uint8_t [] = {
^~~~~~
/node/out/Release/obj/gen/node_code_cache.cc:12:22: error:
decomposition declaration cannot be declared with type 'const uint8_t'
(aka 'const unsigned char'); declared type must be 'auto' or
reference to 'auto'
static const uint8_t [] = {
                     ^
/node/out/Release/obj/gen/node_code_cache.cc:12:22: error:
excess elements in scalar initializer
static const uint8_t [] = {
                     ^
/node/out/Release/obj/gen/node_code_cache.cc:660:7: error:
expected expression
      ,
      ^
/node/out/Release/obj/gen/node_code_cache.cc:661:24: error:
no matching function for call to 'arraysize'
      static_cast<int>(arraysize()), policy
                       ^~~~~~~~~
../src/util.h:667:18: note: candidate function template not viable:
requires 1 argument, but 0 were provided
constexpr size_t arraysize(const T (&)[N]) {
                 ^
2 warnings and 5 errors generated.

This commit suggests that passing a single file be allowed by modifying
tools/js2c.py.

PR-URL: https://github.com/nodejs/node/pull/28443
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

cjihrig committed:
report: include network interfaces in report

PR-URL: https://github.com/nodejs/node/pull/28911
Reviewed-By: Richard Lau <riclau@uk.ibm.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Yongsheng Zhang <zyszys98@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

isaacs committed:
deps: upgrade npm to 6.10.2

PR-URL: https://github.com/nodejs/node/pull/28853
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>
Reviewed-By: Daijiro Wachi <daijiro.wachi@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Yongsheng Zhang <zyszys98@gmail.com>

joaocgreis committed:
test: use unique tmpdirs for each test

Tests can leave processes running blocking the tmpdir. This does not
yet prevent tests from doing that, but prevents failures on
subsequent tests.

PR-URL: https://github.com/nodejs/node/pull/28858
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

gntem committed:
module: add warning when import,export is detected in CJS context

This will allow users to know how to change their project to support
ES modules.

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -85,6 +85,17 @@ const relativeResolveCache = Object.create(null);
+function enrichCJSError(err) {
+  if (
+    err.message.startsWith('SyntaxError: Unexpected token export') ||
+    (/(?:import).*\n{1}\s*\^{9}/mi).test(err.stack)) {
```
Didn't you say you were going to try `err.stack.split('\n')[index]` for this? Or did that not work out?

guybedford replied:
Any update on this?

gntem replied:
Yes, working with splitting the stack by newline and iterating for the line `^^^^^^^^` and getting the previous line, checking if it matches `/import/` should do the trick -- I can guarantee somehow 100% that's precisely detecting it : |

gntem replied:
reworked this a little bit

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -85,6 +85,17 @@ const relativeResolveCache = Object.create(null);
+  if (
+    err.message.startsWith('SyntaxError: Unexpected token export') ||
+    (/(?:import).*\n{1}\s*\^{9}/mi).test(err.stack)) {
+    err.stack = '\nNote: To load an ES module using --experimental-modules ' +
```
Leading `\n` probably not needed here? Or is it?

gntem replied:
This is to separate it from the experimental modules notice and not look line one body

```
(node:27712) ExperimentalWarning: The ESM module loader is experimental.

Note: To load an ES module set "type": "module" in the package.json or use the .mjs extension.
```

guybedford replied:
The warnings are emitted separately to the errors. The same logic would apply to every warning and every error then. So the newline should be removed.

gntem replied:
yes working on this too

gntem replied:
removed this newline from file and test

guybedford replied:
great, thanks.

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -789,36 +800,42 @@ function wrapSafe(filename, content) {
+        '__filename',
+        '__dirname',
+      ]
+    );
```
The try statement must end its wrap here, otherwise we risk intercepting other errors as well, which we shouldn't.

gntem replied:
will work by defining with `let` and function outside the try-catch it the variable is visible to where it's used.

gntem replied:
should be fine now

guybedford commented on test/fixtures/es-modules/invalid.cjs:
```diff
@@ -0,0 +1 @@
+import "invalid";
```
Can you make sure to also include another test for `export`? Maybe a few `export` variations as well.

joaocgreis committed:
fs: close file descriptor of promisified truncate

PR-URL: https://github.com/nodejs/node/pull/28858
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

joaocgreis committed:
repl: close file descriptor of history file

PR-URL: https://github.com/nodejs/node/pull/28858
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

joaocgreis committed:
test: clean tmpdir on process exit

PR-URL: https://github.com/nodejs/node/pull/28858
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

tniessen committed:
test: add test for OAEP hash mismatch

PR-URL: https://github.com/nodejs/node/pull/28335
Fixes: https://github.com/nodejs/node/issues/25756
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Sam Roberts <vieuxtech@gmail.com>

joaocgreis committed:
benchmark: use test/common/tmpdir consistently

Many benchmarks use test/common/tmpdir. This changes 3 benchmarks that
use NODE_TMPDIR to also use test/common/tmpdir.

This is necessary in preparation for the next commit that changes tmpdir
to delete tmpdir.path when the Node.js process exits. Thus, if multiple
benchmarks are run sequentially, the ones that use tmpdir will remove
the directory and the ones changed here would fail because it does not
exist. This happens when running test/benchmark.

Note: to explicitly select a directory for tmpdir, use NODE_TEST_DIR.

PR-URL: https://github.com/nodejs/node/pull/28858
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

himself65 committed:
http2: destroy when settingsFn throws an error

http2.connect should call destroy when init fails.

PR-URL: https://github.com/nodejs/node/pull/28908
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

Trott committed:
doc: remove legacy mode deprecation in assert

The deprecation was revoked in
https://github.com/nodejs/node/pull/28892.

PR-URL: https://github.com/nodejs/node/pull/28909
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Roman Reiss <me@silverwind.io>
Reviewed-By: James M Snell <jasnell@gmail.com>

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -85,6 +85,27 @@ const relativeResolveCache = Object.create(null);
+function enrichCJSError(err) {
+  const stack = err.stack.split('\n');
+
+  const lineWithErr = stack.findIndex((line) =>
```
Is this line not always the same number?

gntem replied:
yes its index `1` i removed `findIndex`

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -819,6 +847,7 @@ function wrapSafe(filename, content) {
   }
 
   return compiled.function;
+
```
This newline should be removed.

gntem replied:
removed

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -85,6 +85,27 @@ const relativeResolveCache = Object.create(null);
+  }
+
+  if (err.message.startsWith('Unexpected token export') ||
+    (previousLine && previousLine.includes('import'))) {
```
This will fail for `var importName = 'invalid;`.

To make it more resilient, we can use the fact that import will always take one of the following forms (if valid syntax):
```
import { ... } from 'x'
import 'x'
import "x"
import x from 'x'
import x, { ... } from 'x'
import * as X from 'x'
```

So if we directly test those variations in a regex we can avoid the invalid cases, which we must do.

gntem replied:
the regexp used now covers most of these cases additionally

```
import {
something
} from "somewhere";
```

but doesn't cover 

```
/* comment */ import "x";
```

i tried to work with the regexp more but i couldn't get it right, any suggestions?

guybedford reviewed (commented):
Great to see the initial copy on the PR itself... we should try and write more here too - something like "this will allow users to know how to change their project to support es modules when they get this error" or similar, try to explain what this does for the user.

himself65 committed:
net: shallow copy option when create Server

PR-URL: https://github.com/nodejs/node/pull/28924
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Masashi Hirano <shisama07@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Сковорода Никита Андреевич <chalkerx@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

mhdawson committed:
doc: remove use of you

We generally avoid the use of 'you'.

PR-URL: https://github.com/nodejs/node/pull/28919
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Beth Griggs <Bethany.Griggs@uk.ibm.com>
Reviewed-By: Richard Lau <riclau@uk.ibm.com>
Reviewed-By: Сковорода Никита Андреевич <chalkerx@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

ChALkeR committed:
Revert "src: remove trace_sync_io_ from env"

This reverts commit 7fa5f54e6f2854183e45e0e105a1e22a381aac60.

The reverted commit breaks the logic behind --trace-sync-io, it should
be enabled only at a certain point in time, while that commit enables
it from the very start, causing warnings be printed for all sync io
instead of sync io after the first tick of the event loop as documented.

Fixes: https://github.com/nodejs/node/issues/28913
Refs: https://github.com/nodejs/node/pull/22726
Refs: https://nodejs.org/api/cli.html#cli_trace_sync_io
PR-URL: https://github.com/nodejs/node/pull/28926
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

reasonablytall committed:
doc: describe NODE_OPTIONS interop w/cmd line opts

Fixes https://github.com/nodejs/node/issues/28910

PR-URL: https://github.com/nodejs/node/pull/28928
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Roman Reiss <me@silverwind.io>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

guybedford commented on test/fixtures/es-modules/unexpected-export-1.cjs:
```diff
@@ -0,0 +1,2 @@
+const example = 10;
+export default example;
```
Please make sure to include separate tests for the cases the regular expression is matching.

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -85,6 +85,20 @@ const relativeResolveCache = Object.create(null);
+  const lineWithErr = stack[1];
+
+  if (err.message.startsWith('Unexpected token export') ||
+    (/^\bimport\b(?=[\s\t\\(\\{\\'\\"\\*\n])?(as)?(from)?/gi).test(lineWithErr)) {
```
1. We don't want the regular expression to be case insensitive because `IMPORT` is not valid import syntax. So we should remove the `i`.
2. `\b` can be replaced with `\s*` I think mostly here
3. I think this is still matching the string - `"import"`, which is wrong, because we specifically don't want to match `import(invalid')`. Perhaps include a test for this.

guybedford reviewed (commented):
Please continue adding tests for these regular expression cases.

It's fine if the regular expression doesn't capture eg comments (that is uses of module syntax we don't catch can be ok).

What we need to be careful about is to very specifically ensure there are no valid scripts that will give this module warning!! And that is the thing we test with eg `import(sdf')` testing.

Trott committed:
test: refactor test-fs-stat.js

General code cleanup.

PR-URL: https://github.com/nodejs/node/pull/28929
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

guybedford committed:
module: pkg exports validations and fallbacks

PR-URL: https://github.com/nodejs/node/pull/28949
Reviewed-By: Bradley Farias <bradley.meck@gmail.com>
Reviewed-By: Jan Krems <jan.krems@gmail.com>

vsemozhetbyt committed:
doc: fix sorting nit in sections of http.md

PR-URL: https://github.com/nodejs/node/pull/28943
Reviewed-By: Richard Lau <riclau@uk.ibm.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>

guybedford commented on test/fixtures/es-modules/es-note-unexpected-import-5.cjs:
```diff
@@ -0,0 +1 @@
+        import 'here';
```
The test with spaces to check was `import     ('invalid')`.

guybedford commented on lib/internal/modules/cjs/loader.js:
```diff
@@ -85,6 +85,20 @@ const relativeResolveCache = Object.create(null);
+  const lineWithErr = stack[1];
+
+  if (err.message.startsWith('Unexpected token export') ||
+    (/^\s*import(?=[ {'"*])\s*/).test(lineWithErr)) {
```
Should just be `/^\s*import(?=[ {'"*])\s*(?![ (])/` actually. By including a space in the negative capture group we can avoid the backtracking.

mscdex committed:
path: improve normalization performance

PR-URL: https://github.com/nodejs/node/pull/28948
Reviewed-By: Yongsheng Zhang <zyszys98@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Gus Caplan <me@gus.host>

addaleax committed:
events: give subclass name in unhandled 'error' message

For unhandled `'error'` events, include the constructor name for
subclasses of EventEmitter, if possible. This makes tracing errors
easier when both creation of the `Error` object and emitting it
happen in code that does not refer back to the event emitter.

PR-URL: https://github.com/nodejs/node/pull/28952
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

mscdex committed:
util: improve debuglog performance

PR-URL: https://github.com/nodejs/node/pull/28956
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Anto Aravinth <anto.aravinth.cse@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Joyee Cheung <joyeec9h3@gmail.com>
Reviewed-By: Gus Caplan <me@gus.host>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>

ronag committed:
fs: fix (temporary) for esm package

Fix to unblock CITGM. See,
https://github.com/standard-things/esm/issues/821.

PR-URL: https://github.com/nodejs/node/pull/28957
Reviewed-By: Ruben Bridgewater <ruben@bridgewater.de>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>

EduardoRFS committed:
doc: make unshift doc compliant with push doc

readable.unshift() also allows to pass null and end stream

PR-URL: https://github.com/nodejs/node/pull/28953
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

mscdex committed:
stream: improve read() performance

PR-URL: https://github.com/nodejs/node/pull/28961
Reviewed-By: Weijia Wang <starkwang@126.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Yongsheng Zhang <zyszys98@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>

cjihrig committed:
report: list envvars using uv_os_environ()

This commit simplifies the diagnostic report's code for listing
environment variables by using uv_os_environ() instead of
platform specific code.

PR-URL: https://github.com/nodejs/node/pull/28963
Reviewed-By: Saúl Ibarra Corretgé <saghul@gmail.com>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>

guybedford committed:
module: refine package name validation

PR-URL: https://github.com/nodejs/node/pull/28965
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Jan Krems <jan.krems@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

ronag committed:
stream: encapsulate buffer-list

PR-URL: https://github.com/nodejs/node/pull/28974
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

cjihrig committed:
fs: validate fds as int32s

This commit updates the JS layer's validation of file
descriptors to check for int32s >= 0 instead of uint32s.

PR-URL: https://github.com/nodejs/node/pull/28984
Fixes: https://github.com/nodejs/node/issues/28980
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Gus Caplan <me@gus.host>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

nfearnley committed:
doc: documented default statusCode

PR-URL: https://github.com/nodejs/node/pull/28982
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

mscdex committed:
benchmark: allow easy passing of process flags

PR-URL: https://github.com/nodejs/node/pull/28986
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Anatoli Papirovski <apapirovski@mac.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>

mscdex committed:
stream: improve read() performance more

PR-URL: https://github.com/nodejs/node/pull/28989
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>

BridgeAR committed:
2019-08-06, Version 12.8.0 (Current)

Notable changes:

* assert:
  * Legacy mode deprecation (`DEP0089`) is revoked (Colin Ihrig)
    https://github.com/nodejs/node/pull/28892
* crypto:
  * The `outputLength` option is added to `crypto.createHash`
    (Tobias Nießen) https://github.com/nodejs/node/pull/28805
  * The `maxmem` range is increased from 32 to 53 bits (Tobias Nießen)
    https://github.com/nodejs/node/pull/28799
* n-api:
  * Added APIs for per-instance state management (Gabriel Schulhof)
    https://github.com/nodejs/node/pull/28682
* report:
  * Network interfaces get included in the report (cjihrig)
    https://github.com/nodejs/node/pull/28911
* src:
  * `v8.getHeapCodeStatistics()` is now exported
    (Yuriy Vasiyarov) https://github.com/nodejs/node/pull/27978

PR-URL: https://github.com/nodejs/node/pull/29017

ronag committed:
http: add missing stream-like properties to OutgoingMessage

PR-URL: https://github.com/nodejs/node/pull/29018
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

ronag committed:
http: buffer writes even while no socket assigned

PR-URL: https://github.com/nodejs/node/pull/29019
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>

addaleax committed:
test: refactor test-sync-io-option

Refactor `test/parallel/test-sync-io-option.js` to be simpler and cover
more cases (in particular, this adds a regression test for
https://github.com/nodejs/node/issues/28913).

Refs: https://github.com/nodejs/node/issues/28913

PR-URL: https://github.com/nodejs/node/pull/29020
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

BethGriggs committed:
2019-08-06, Version 10.16.2 'Dubnium' (LTS)

Notable changes:

This release patches a regression in the OpenSSL upgrade to 1.1.1c that
 causes intermittent hangs in machines that have low entropy.

Refs: https://github.com/nodejs/node/issues/28932

PR-URL: https://github.com/nodejs/node/pull/29021

ronag committed:
http: remove duplicate check

`socketOnDrain()` performs the exact same check and doesn't return
anything.

PR-URL: https://github.com/nodejs/node/pull/29022
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

vsemozhetbyt committed:
doc: fix nits in child_process.md

* Use `console.error()` for error or stderr output.
* Unify comment style.
* Unify link format.
* Correct link URL.
* Fix some typos.

PR-URL: https://github.com/nodejs/node/pull/29024
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Anto Aravinth <anto.aravinth.cse@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

ljharb committed:
test: assert: add failing deepEqual test for faked boxed primitives

PR-URL: https://github.com/nodejs/node/pull/29029
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Gus Caplan <me@gus.host>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Rich Trott <rtrott@gmail.com>

ljharb committed:
util: assert: fix deepEqual comparing fake-boxed to real boxed primitive

PR-URL: https://github.com/nodejs/node/pull/29029
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Gus Caplan <me@gus.host>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Rich Trott <rtrott@gmail.com>

ljharb committed:
util: isEqualBoxedPrimitive: ensure both values are actual boxed Symbols

... before trying to valueOf them

PR-URL: https://github.com/nodejs/node/pull/29029
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Gus Caplan <me@gus.host>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Rich Trott <rtrott@gmail.com>

tniessen committed:
doc: improve UV_THREADPOOL_SIZE description

PR-URL: https://github.com/nodejs/node/pull/29033
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Anto Aravinth <anto.aravinth.cse@gmail.com>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Beth Griggs <Bethany.Griggs@uk.ibm.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

ronag committed:
stream: inline and simplify onwritedrain

Inline and simplify onwritedrain. Also remove comment that seems to be
outdated/invalid.

PR-URL: https://github.com/nodejs/node/pull/29037
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Matteo Collina <matteo.collina@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>

cjihrig committed:
dns: update lookupService() first arg name

The first argument to lookupService() should be an IP address,
and is named "address" in the documentation. This commit updates
the code to match the documentation and provide less confusing
errors.

PR-URL: https://github.com/nodejs/node/pull/29040
Fixes: https://github.com/nodejs/node/issues/29039
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Sam Roberts <vieuxtech@gmail.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Michaël Zasso <targos@protonmail.com>
Reviewed-By: James M Snell <jasnell@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>

sam-github committed:
test: unskip tests that now pass on AIX

One skipped test remains, it creates very large Buffer objects,
triggering the AIX OOM to kill node and its parent processes.

See: https://github.com/nodejs/build/issues/1849#issuecomment-514414165

PR-URL: https://github.com/nodejs/node/pull/29054
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Luigi Pinca <luigipinca@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Richard Lau <riclau@uk.ibm.com>

cjihrig committed:
deps: upgrade to libuv 1.31.0

Notable changes:

- UV_FS_O_FILEMAP has been added for faster access to memory
  mapped files on Windows.
- uv_fs_mkdir() now returns UV_EINVAL for invalid filenames
  on Windows. It previously returned UV_ENOENT.
- The uv_fs_statfs() API has been added.
- The uv_os_environ() and uv_os_free_environ() APIs have
  been added.

Fixes: https://github.com/nodejs/node/issues/28599
Fixes: https://github.com/nodejs/node/issues/28945
Fixes: https://github.com/nodejs/node/issues/29008
PR-URL: https://github.com/nodejs/node/pull/29070
Reviewed-By: Richard Lau <riclau@uk.ibm.com>
Reviewed-By: Trivikram Kamat <trivikr.dev@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Ben Noordhuis <info@bnoordhuis.nl>
Reviewed-By: Jiawen Geng <technicalcute@gmail.com>

Commit:
src: organize imports in inspector_profiler.cc

In the other .cc files in the project, includes are in alphabetical
order, with local files first, and libraries after. However,
inspector_profiler.cc has a library declared in the middle of the import
order, and v8 is the second to last being imported, instead of the last.
So I reordered the imports and testing showed no side effects;
everything passed.

PR-URL: https://github.com/nodejs/node/pull/29073
Reviewed-By: Anna Henningsen <anna@addaleax.net>
Reviewed-By: Eugene Ostroukhov <eostroukhov@google.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
Reviewed-By: Yongsheng Zhang <zyszys98@gmail.com>

lpinca committed:
http: remove redundant condition

`conn.destroyed` is guaranteed to be `false` because a previous `if`
statement already handles the case where `conn && conn.destroyed`
evaluates to `true` returning `false` in that case.

PR-URL: https://github.com/nodejs/node/pull/29078
Reviewed-By: Colin Ihrig <cjihrig@gmail.com>
Reviewed-By: Rich Trott <rtrott@gmail.com>
